circuit_knitting/cutting/cut_finding/cco_utils.py

```diff
@@ -138,23 +138,23 @@ def greedy_best_first_search(
         Callable, search_space_funcs.next_state_func
     )
 
-    if search_space_funcs.goal_state_func(state, *args):
-        return state
-
-    best = min(
-        [
-            (search_space_funcs.cost_func(next_state, *args), k, next_state)
-            for k, next_state in enumerate(
-                search_space_funcs.next_state_func(state, *args)
-            )
-        ],
-        default=(None, None, None),
-    )
-
-    if best[-1] is None:  # pragma: no cover
-        # This covers a rare edge case.
-        # We have so far found no circuit that triggers it.
-        # Excluding from test coverage for now.
-        return None
-
-    return greedy_best_first_search(best[-1], search_space_funcs, *args)
+    while not search_space_funcs.goal_state_func(state, *args):
+        best = min(
+            [
+                (search_space_funcs.cost_func(next_state, *args), k, next_state)
+                for k, next_state in enumerate(
+                    search_space_funcs.next_state_func(state, *args)
+                )
+            ],
+            default=(None, None, None),
+        )
+        if best[-1] is None:  # pragma: no cover
+            # This covers a rare edge case.
+            # We have so far found no circuit that triggers it.
+            # Excluding from test coverage for now.
+            return None
+        else:
+            # avoid using recursion to stay clear of any risk of stack overflow.
+            state = best[-1]
+            continue
```

Comment:
Following [Caleb's comment](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/pull/537/files#r1557810322), I suggest

```suggestion
        state = best[-1]
```

Reply:
Thanks, I have made this change here: [183c13a](https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/pull/537/commits/183c13a2d91fc9b40b45f43db61b1106d6eb9d8e)